<a href="https://colab.research.google.com/github/cyrus723/my-first-binder/blob/main/sentiment_analysis_1_all_models.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [23]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# transformer

In [ ]:
from transformers import pipeline

def sentiment_analysis1(text):
  # Load the sentiment analysis pipeline with the BERT model
  classifier = pipeline("sentiment-analysis", model="nlptown/bert-base-multilingual-uncased-sentiment")

  # Perform sentiment analysis
  result = classifier(text)
  return result

# Example usage:
financial_news = "The stock market experienced a significant downturn today amid concerns over inflation."
result = sentiment_analysis1(financial_news)
print("Sentiment Analysis Result:", result)


Sentiment Analysis Result: [{'label': '2 stars', 'score': 0.47418859601020813}]


Label Interpretation: The label '2 stars' suggests that the sentiment expressed in the text you analyzed is generally negative. In sentiment rating systems where labels are presented as stars (often ranging from 1 star to 5 stars), a rating of 2 stars typically denotes dissatisfaction or a negative view. It implies that the text likely contains criticisms or less favorable opinions.

Score Interpretation: The confidence score of approximately 0.474 (or 47.4%) associated with this label indicates a moderate level of confidence in the assessment. This isn't a very high confidence level, which might suggest that the sentiment expressed in the text was not overwhelmingly clear or that the text contained mixed sentiments, making it harder for the model to assign a more definitive sentiment rating with higher confidence.

Considerations:

In [ ]:
data = {
    'headline': [
         "Apple's revenue exceeds expectations with a strong quarterly report",
         "Major banks face scrutiny as financial regulations tighten",
         "Tesla's new factory investment raises concerns over debt levels"
     ]
}

df = pd.DataFrame(data)
df

headline
0  Apple's revenue exceeds expectations with a st...
1  Major banks face scrutiny as financial regulat...
2  Tesla's new factory investment raises concerns...

In [ ]:
df['sentiment'] = sentiment_analysis1(df['headline'].tolist())
print(df)

                                            headline  \
0  Apple's revenue exceeds expectations with a st...   
1  Major banks face scrutiny as financial regulat...   
2  Tesla's new factory investment raises concerns...   

                                           sentiment  
0  {'label': '5 stars', 'score': 0.5569013357162476}  
1  {'label': '4 stars', 'score': 0.27764415740966...  
2  {'label': '5 stars', 'score': 0.4056451618671417}  


# FinBERT

We will load a FINBERT model from Hugging Face’s model repository. Please note that, as of my last training data, specific models like ‘yiyanghkust/finbert-tone’ can be used for demonstration:

In [ ]:
from transformers import pipeline

# Initialize the sentiment analysis model
model_name = "yiyanghkust/finbert-tone"
finbert = pipeline("sentiment-analysis", model=model_name)

# Function to analyze sentiment
def analyze_sentiment(text):
    results = finbert(text)
    return results

# Example usage:
financial_news = "The company's revenue has exceeded expectations, but concerns about regulatory challenges remain."
sentiment_result = analyze_sentiment(financial_news)

# Print the results
print("Sentiment Analysis Result:", sentiment_result)


Sentiment Analysis Result: [{'label': 'Positive', 'score': 0.9999998807907104}]


In [ ]:
df['sentiment'] = analyze_sentiment(df['headline'].tolist())
print(df)

                                            headline  \
0  Apple's revenue exceeds expectations with a st...   
1  Major banks face scrutiny as financial regulat...   
2  Tesla's new factory investment raises concerns...   

                                           sentiment  
0  {'label': 'Positive', 'score': 0.9999991655349...  
1  {'label': 'Negative', 'score': 0.9769049882888...  
2  {'label': 'Negative', 'score': 0.9999909400939...  


In [1]:
pos="In two years at Texas State Dr. Yi was perhaps the most exceptional teacher \
I have had. His passion for what he teaches as well as his students well-being is \
evident. He went far beyond what is necessary to provide us with material via his canvas pages. \
I would highly highly recommend other students in Finance take his course(s) they would be better for it. \
Thank you"

In [2]:
neg ="Was hard to understand what the professor was saying. \
Materials on the test sometimes would reflect only a small fraction of the homework and review,\
the rest seemed to be vague in where he got the information from. \
It was also hard to stay engaged during class, I fell asleep every time I attended class. \
The lectures were practically useless to me"

### Using TextBlob

In [5]:
#!pip install textblob

from textblob import TextBlob
print(TextBlob(pos).sentiment)
print(TextBlob(neg).sentiment)

Sentiment(polarity=0.25645833333333334, subjectivity=0.645625)
Sentiment(polarity=-0.3055555555555556, subjectivity=0.5305555555555556)


In [ ]:
!pip install vaderSentiment

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 125 kB 5.4 MB/s 


In [19]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
vs_pos = SentimentIntensityAnalyzer().polarity_scores(pos)
vs_neg = SentimentIntensityAnalyzer().polarity_scores(neg)

print("{:-<65} {}".format(pos, str(vs_pos)))
print('--------------')
print("{:-<65} {}".format(neg, str(vs_neg)))
print(vs_pos)
print(vs_neg)

In two years at Texas State Dr. Yi was perhaps the most exceptional teacher I have had. His passion for what he teaches as well as his students well-being is evident. He went far beyond what is necessary to provide us with material via his canvas pages. I would highly highly recommend other students in Finance take his course(s) they would be better for it. Thank you {'neg': 0.0, 'neu': 0.82, 'pos': 0.18, 'compound': 0.9113}
--------------
Was hard to understand what the professor was saying. Materials on the test sometimes would reflect only a small fraction of the homework and review,the rest seemed to be vague in where he got the information from. It was also hard to stay engaged during class, I fell asleep every time I attended class. The lectures were practically useless to me {'neg': 0.107, 'neu': 0.852, 'pos': 0.041, 'compound': -0.3182}
{'neg': 0.0, 'neu': 0.82, 'pos': 0.18, 'compound': 0.9113}
{'neg': 0.107, 'neu': 0.852, 'pos': 0.041, 'compound': -0.3182}


In [22]:
!pip install flair

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 388.3/388.3 kB 6.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.3/139.3 kB 8.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.4/54.4 kB 6.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.7/19.7 MB 40.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 981.5/981.5 kB 60.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 202.6/202.6 kB 24.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.8/143.8 kB 15.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.2/12.2 MB 49.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.2/82.2 kB 10.6 MB/s eta 0:00:00
  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-

In [26]:
from flair.models import TextClassifier
from flair.data import Sentence

classifier = TextClassifier.load('en-sentiment')

sentence = Sentence(pos)
classifier.predict(sentence)

# print sentence with predicted labels
print('Sentence above is: ', sentence.labels)

sentence = Sentence(neg)
classifier.predict(sentence)

# print sentence with predicted labels
print('Sentence above is: ', sentence.labels)

Sentence above is:  ['Sentence[74]: "In two years at Texas State Dr. Yi was perhaps the most exceptional teacher I have had. His passion for what he teaches as well as his students well-being is evident. He went far beyond what is necessary to provide us with material via his canvas pages. I would highly highly recommend other students in Finance take his course(s) they would be better for it. Thank you"'/'POSITIVE' (0.9975)]
Sentence above is:  ['Sentence[65]: "Was hard to understand what the professor was saying. Materials on the test sometimes would reflect only a small fraction of the homework and review,the rest seemed to be vague in where he got the information from. It was also hard to stay engaged during class, I fell asleep every time I attended class. The lectures were practically useless to me"'/'NEGATIVE' (1.0)]


In [28]:
from transformers import pipeline

def sentiment_analysis1(text):
  # Load the sentiment analysis pipeline with the BERT model
  classifier = pipeline("sentiment-analysis", model="nlptown/bert-base-multilingual-uncased-sentiment")

  # Perform sentiment analysis
  result = classifier(text)
  return result

# Example usage:

result = sentiment_analysis1(pos)
print("Sentiment Analysis Result:", result)

result = sentiment_analysis1(neg)
print("Sentiment Analysis Result:", result)


Sentiment Analysis Result: [{'label': '5 stars', 'score': 0.7873240113258362}]
Sentiment Analysis Result: [{'label': '2 stars', 'score': 0.6215679049491882}]


In [30]:
from transformers import pipeline

# Initialize the sentiment analysis model
model_name = "yiyanghkust/finbert-tone"
finbert = pipeline("sentiment-analysis", model=model_name)

# Function to analyze sentiment
def analyze_sentiment(text):
    results = finbert(text)
    return results

# Example usage:
sentiment_result = analyze_sentiment(pos)

# Print the results
print("Sentiment Analysis Result:", sentiment_result)


# Example usage:
sentiment_result = analyze_sentiment(neg)

# Print the results
print("Sentiment Analysis Result:", sentiment_result)


Sentiment Analysis Result: [{'label': 'Positive', 'score': 0.9995104074478149}]
Sentiment Analysis Result: [{'label': 'Neutral', 'score': 0.999244213104248}]
